src/main/java/io/cryostat/diagnostic/DiagnosticsHelper.java

```diff
@@ -0,0 +1,341 @@
+
+    @Inject
+    @Named(Producers.BASE64_URL)
+    Base64 base64Url;
+
+    @Inject S3Client storage;
+    @Inject Logger log;
+
+    @Inject Instance<BucketedDiagnosticsMetadataService> metadataService;
+
+    private static final String DUMP_THREADS = "threadPrint";
+    private static final String DUMP_THREADS_TO_FIlE = "threadDumpToFile";
+    private static final String DIAGNOSTIC_BEAN_NAME = "com.sun.management:type=DiagnosticCommand";
+    static final String THREAD_DUMP_REQUESTED = "ThreadDumpRequested";
+    static final String THREAD_DUMP_DELETED = "ThreadDumpDeleted";
+    private static final String META_KEY_NAME = "uuid";
+    private static final String META_KEY_JVMID = "jvmId";
+
+    public static final String METADATA_STORAGE_MODE_TAGGING = "tagging";
+    public static final String METADATA_STORAGE_MODE_OBJECTMETA = "metadata";
+    public static final String METADATA_STORAGE_MODE_BUCKET = "bucket";
+
+    @Inject EventBus bus;
+    @Inject TargetConnectionManager targetConnectionManager;
+    @Inject StorageBuckets buckets;
+
+    void onStart(@Observes StartupEvent evt) {
+        log.tracev("Creating thread dump bucket: {0}", bucket);
+        buckets.createIfNecessary(bucket);
+    }
+
+    public ThreadDump dumpThreads(String format, long targetId) {
+        if (!(format.equals(DUMP_THREADS) || format.equals(DUMP_THREADS_TO_FIlE))) {
+            throw new BadRequestException();
+        }
+        log.tracev(
+                "Thread Dump request received for Target: {0} with format: {1}", targetId, format);
+        Object[] params = new Object[1];
+        String[] signature = new String[] {String[].class.getName()};
+        return targetConnectionManager.executeConnectedTask(
+                Target.getTargetById(targetId),
+                conn -> {
+                    String content =
+                            conn.invokeMBeanOperation(
+                                    DIAGNOSTIC_BEAN_NAME, format, params, signature, String.class);
+                    return addThreadDump(content, Target.getTargetById(targetId).jvmId);
+                });
+    }
+
+    public List<ThreadDump> getThreadDumps(long targetId) {
+        return listThreadDumps(targetId).stream()
+                .map(
+                        item -> {
+                            try {
+                                return convertObject(item);
+                            } catch (Exception e) {
+                                log.error(e);
+                                return null;
+                            }
+                        })
+                .filter(
+                        item -> {
+                            log.tracev("Item jvmID: {0}", item.jvmId());
+                            log.tracev("Item key: {0}", item.uuid());
+                            log.tracev("Item download URL: {0}", item.downloadUrl());
+                            return Target.<Target>findById(targetId).jvmId.equals(item.jvmId());
```

Comment:
There's an NPE risk here since `Target.jvmId` is nullable, in particular in the case where a target has been discovered but Cryostat has not yet connected to it to retrieve the JVM ID. That might simply be due to timing, or it might be because Cryostat does not trust the JVM's JMX server cert or does not have appropriate stored credentials. So this would be safer as ex. `Objects.equals(Target.<Target>findById(targetId).jvmId, item.jvmId())`.

Also, `Target.<Target>findById(targetId)` will be null if the given `targetId` doesn't correspond to anything, so that's another NPE risk.

Reply:
`Target.getTargetById()` could be used instead of `Target.<Target>findById()`, since it also returns a `Target` but will throw if no such ID exists (or if somehow more than one entity has the same ID, which should be impossible). At least that way it's a more specific exception that can be handled better.